### DISCUSSION TITLE
Why is a private variable explicit getter more gas efficient than the public auto-generated one?

### DISCUSSION THREAD

**Discussion by iliyan03** [Cyfrin/foundry-full-course-cu#1861] [Q&A]
Hi everyone 👋
I am currently working on the smart contract lottery project. As I was creating the first variable and its getter, one question came to my mind: "Is this more efficient than making the variable public and using its automatically generated getter?". So I did a test. I put the code into the Remix IDE, compiled it and called the explicit getter of the private property. The gas cost was 310.
Then I replaced the "private" keyword with "public", removed the explicit getter, recompiled and called the property directly to get its value. Here the gas was 347.
And now I'm wondering why the automatically generated getter is more gas inefficient than the explicit one? Can anyone help me understand this?

Here's the code with a private property and explicit getter:
```solidity
contract Raffle {
    uint256 private immutable i_entranceFee;

    constructor(uint256 entranceFee){
        i_entranceFee = entranceFee;
    }

    function getEntranceFee() external view returns (uint256) {
        return i_entranceFee;
    }
}
```

And that's with a public property:
```solidity
contract Raffle {
    uint256 public immutable i_entranceFee;

    constructor(uint256 entranceFee){
        i_entranceFee = entranceFee;
    }
}
```

**Comment by EngrPips** [comment]
```js
contract Raffle {
uint256 private immutable i_entranceFee;

constructor(uint256 entranceFee){
    i_entranceFee = entranceFee;
}

function enterRaffle() public payable {}

function pickWinner() public {}

function getEntranceFee() external view returns (uint256) {
    return i_entranceFee;
}
}
```

```js
contract Raffle {
uint256 public immutable i_entranceFee;

constructor(uint256 entranceFee){
    i_entranceFee = entranceFee;
}

function enterRaffle() public payable {}

function pickWinner() public {}
}
```

**Reply by EngrPips** [reply]
Hello @iliyan03, won't you mind re-doing that experiment but making the variable not immutable? Let's see what the gas output would be for the two processes. Normally, any immutable or constant variable is always a part of the contract bytecode, so reading them normally would be cheaper. I did an experiment and some research myself and saw that I got it wrong. But here is a link to a medium post `https://medium.com/coinmonks/getter-and-setter-functions-in-solidity-67a3a2cb93c` that explained what could be the cause of why `automatically generated getter` is more costly than an explicit getter. I think solidity do some extra stuffs for `automatically generated getter` which we normally don't do in our explicit getter.

**Reply by iliyan03** [reply]
Sure, I did the experiment again.
- not immutable, private variable, explicit getter: gas cost = 2415
- not immutable, public variable: gas cost = 2451
It is indeed much more gas efficient to make the variables immutable or constant. But the difference in gas cost is almost the same.
Also, how does making the variables private or public affect the security of the contract? Don't these visibility keywords just provide a better and more gas-efficient interface?

**Reply by EngrPips** [reply]
Interesting. Personally, I get into the habit of making all of my variables private and creating a getter for them if there is a need for it, and this has helped in marking my functions also appropriately and not opening an attack path where I shouldn't in my contract, For me it is just some sort of security mindset, and actually it cost less gas.

**Reply by iliyan03** [reply]
Pretty much the same thing here - making everything private and change it later if there's a need. Much better interface and is more gas efficient.
BTW, you can check @thebraveman17's reply, it sounds like it's the answer to my question, but I did a test where I made the method `public` instead of `external`, but the gas cost was still the same. If you have any idea why this is happening, I would be grateful if you could share it with us.

**Reply by EngrPips** [reply]
You know, previously, there was always this mentioning that `external` function consumes less gas than `public` function, but over time, it became evident that it isn't actually genuine. Honestly, for your question, what I would say is explicitly defining a getter limits the process the contract would go through before it gets the variable we want because we explicitly tell the contract what to do to get the variable, but making a variable public is us telling the contract do whatever you need to do to get us this variable and that might include some extra step which we wouldn't do when we explicitly define our getter. Finally, remember that for an `EVM` machine, every step and every opcode costs gas, so more steps would lead to more gas costs. I should read up more and ask people who know more to learn in detail what processes are included when we get variables through an `automatically generated getter` that are excluded when we use `explicitly defined getters` that make `automatically generated getter` cost more gas.

**Reply by iliyan03** [reply]
Great answer, @EngrPips! Thanks! Would appreciate you getting some more info on the topic. I did compare the opcodes of both versions, but I'm not on that level yet to understand what's actually going on :D

**Reply by EngrPips** [reply]
I accept the proposal to research it and discover what is happening under the hood. I will write a simple solidity contract that gets a variable using an `automatically generated getter` and one that gets a variable using `explicitly defined getters.` I will compare the opcode of both getters and see what they are doing differently. I will come back here with my findings.

<img width="582" alt="Screenshot 2024-06-20 at 15 58 08" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/95760918/cc510910-56c1-4303-91a9-5e57f3f928c8">

Thanks for the push to learn @iliyan03.

**Reply by iliyan03** [reply]
No worries, brother! That's what we're here for, to help each other and maybe learn something new. Can't wait to see what you come up with!

**Comment by urosognjenovic** [comment]
The variable and its automatically generated getter are ```public```, while the explicit getter function is ```external```. ```External``` functions consume less gas than their ```public``` counterparts.

**Reply by iliyan03** [reply]
This makes sense, but making the explicit getter a `public` method didn't change anything - the gas cost was still the same.

**Reply by urosognjenovic** [reply]
You're right, I also tested this. 

This made me dig a bit about ```external``` and ```public``` functions and I found this:
https://ethereum.stackexchange.com/questions/19380/external-vs-public-best-practices
https://ethereum.stackexchange.com/questions/147417/solidity-public-or-external-function-which-is-more-gas-efficient

As it seems, ```public``` and ```external``` functions consume the same amount of gas since solidity 0.6.9. Before that, ```public``` functions would automatically copy the arguments to memory, as opposed to ```external``` functions which allowed the use of calldata.

So to correct myself, ```private``` variables consume less gas than ```public``` variables. Given that you use a lot of variables in your contract, you make them private by default to save gas. If you need to access those variables from outside the contract, you make an explicit getter function. I didn't analyze the exact opcodes that are used by ```private``` and ```public``` variables to find the difference. Maybe I will do it some day when I have enough spare time.

I hope this is clear now.

**Reply by iliyan03** [reply]
Thank you, @thebraveman17! I appreciate the effort and resources you have shared!
If the question is still unanswered and you have some time to look into it, I'd be grateful to see what you've found out about the cause of the difference in gas costs.

**Reply by xueyuanhuang** [reply]
hey @iliyan03  I have similar question about keyword public compared to private.
You know in video, sometimes we use `public` keywords, sometimes we use `private`, I am quite confused about wen to use which.

I ask chatgpt, it said pretty much you guys said. Maybe I need to dig more, as gas efficience can't persuade me yet. Just leave a question mark here.

Thanks for the question!

**Reply by EngrPips** [reply]
Interesting.
